### DISCUSSION TITLE
Clear Commit History in GitHUb for Desktop

### DISCUSSION THREAD

**Discussion by ksammons** [community/community#33261] [New to GitHub]
Hi Everyone,

My team and I are using GitHUb for Desktop and are struggling with random commits showing up on clean branches. For example, we will make a new branch from master, work on some changes, and when we go to make a PR there will be commits from other users on our branch.

This happens no matter what--even if we rebase with master before we create the PR. Is this because we need to clear our commit history? If so, is there a way to do this via the desktop app? None of us are very familiar with how to use command line and are sort of hobbling along without much knowledge.

Thank you so much for your help!

**Comment by airtower-luna** [comment]
> when we go to make a PR there will be commits from other users on our branch.

What exactly do you mean by that?

**Reply by ksammons** [reply]
Let me try and explain better :)

Say Bob makes a new branch off master. He does some work/makes some commits and wants to push back up to master. Let's say Bob changed two files. Bob makes the PR but the PR says it has 110 changed files with commits from Steve, Mike, and Bill. In some cases these commits say they are from months ago. Even if the branch was made from master that day and no other changes have been pushed to master in the meantime.

I think we are making branches off incorrect commits, but I don't know why that is happening when we are pulling origin/updating our repos all the time.

**Reply by airtower-luna** [reply]
Definitely sounds like it, but hard to say what exactly happens without seeing your repository. The most likely reason in my experience would be that you merge PRs using the "rebase" or "squash" methods. That messes with history, either by recreating commits (rebase) which changes their IDs, or creating completely new commits (squash), even if the resulting code is the same.

If someone then has the original commits on their brach those are different commits, so a merge would bring them along again.

**Reply by ksammons** [reply]
Okay, so it definitely sounds like when need to clean our history...somehow. Right now the devs have us using cherry-picking into the lesser environments, and I don't think anyone on my team knows how to squash. But that doesn't mean one of the devs/other people using our repo aren't doing that. We are super new to this and I appreciate your help. :)

**Reply by airtower-luna** [reply]
Cherry-picking would explain what you see, too. It effectively recreates a commit: Same diff (but different resulting tree) and same (or similar, if adding a "cherry-picked from" line) commit message, different ancestor, different commit time. Any of those changes means the cherry-picked commit gets a new ID. And if you merge the branches later that means the original and cherry-picked commits are different commits which will both be in the merged history. :slightly_smiling_face: 

> I don't think anyone on my team knows how to squash.

Keep it that way. In my projects I prohibit squash merges completely, because they mess up history so much. :wink: 

In general, if you're new to Git I recommend you check the [Git Book](https://git-scm.com/book/en/v2) website. Chapter 3 gives a pretty good overview of working with branches. At work I tell new colleagues who aren't very familiar with Git yet to read at least chapters 2 and 3.

**Reply by ksammons** [reply]
Thank you so very much!!!!!! I will share this with our devs/my team and see if we can find a solution. Much appreciated. :)
